code-studio/src/js/components/GridEditor.jsx

```diff
@@ -0,0 +1,176 @@
+        cells = this.state.serializedData.map(function (row) {
+          return row.map(cellClass.deserialize);
+        });
+      } else if (this.props.serializedMaze) {
+        cells = this.props.serializedMaze.map(function (row) {
+          return row.map(cellClass.deserialize);
+        });
+      } else {
+        cells = this.props.maze.map(function (row, x) {
+          return row.map(function (mazeCell, y) {
+            var initialDirtCell = this.props.initialDirt[x][y];
+            return cellClass.parseFromOldValues(mazeCell, initialDirtCell);
+          }, this);
+        }, this);
+      }
+
+      return cells;
+    },
+
+    render: function () {
+
+      var cells = this.getDeserializedCells();
+
+      var rows = cells.map(function (row, x) {
+        var cells = row.map(function (cell, y) {
+          var classNames = [];
+          var tdStyle = {};
+
+          var tiles = ['border', 'path', 'start', 'end', 'obstacle'];
+          classNames.push(tiles[cell.tileType_]);
+
+          if (mazeUtils.isBeeSkin(this.props.skin)) {
+            var conditions = ['', 'flower-or-hive', 'flower-or-nothing', 'hive-or-nothing', 'flower-hive-or-nothing'];
+            var features = ['hive', 'flower'];
+            if (cell.isVariableCloud()) {
+              classNames.push('conditional');
+              classNames.push(conditions[cell.cloudType_]);
+            } else if (cell.featureType_ !== undefined) {
+              classNames.push(features[cell.featureType_]);
+            }
+          } else {
+            // farmer
+            if (cell.isDirt()) {
+              classNames.push('dirt');
+              var dirtValue = cell.getCurrentValue();
+              var dirtIndex = 10 + dirtValue + (dirtValue < 0 ? 1 : 0);
+              tdStyle.backgroundPosition = -dirtIndex * 50;
+            }
+          }
+
+          var text = '';
+          if (cell.originalValue_ !== undefined && cell.originalValue_ !== null) {
+            text = cell.originalValue_.toString();
+            if (cell.range_ && cell.range_ > cell.originalValue_) {
+              text += " - " + cell.range_.toString();
+            }
+          }
+
+          return (<td data-row={x} data-col={y} className={classNames.join(' ')} onClick={this.clickCell} style={tdStyle}>
+            {text}
+          </td>);
+        }, this);
+        return (<tr>
+          {cells}
+        </tr>);
+      }, this);
+
+      var cellEditor;
+      var selectedCellJson;
+      var row = this.state.selectedRow;
+      var col = this.state.selectedCol;
+      if (cells[row] && cells[row][col]) {
+        var cell = cells[row][col];
+        cellEditor = mazeUtils.isBeeSkin(this.props.skin) ?
+            <BeeCellEditor cell={cell} row={row} col={col} onUpdate={this.onCellChange} /> :
+            <CellEditor cell={cell} row={row} col={col} onUpdate={this.onCellChange} />;
+
+        selectedCellJson = (<label>
+          Cell JSON (for copy/pasting):
```

Comment:
It's for levelbuilders; copy/paste support was the one grid editing feature that was specifically requested.

I debated a couple different ways to try to hide it, but they all ended up feeling a bit weird. Once the levelbuilders get their hands on this, I'll find out from them if it's confusing or not as it stands and hopefully not have to do anything.